### PR TITLE
Refactor query status polling logic in EMRContainerHook

### LINKED ISSUE #19877: Refactor poll_query_status in EMRContainerHook
### Body

The goal is to refactor the code so that we can remove this TODO
https://github.com/apache/airflow/blob/7640ba4e8ee239d6e2bbf950d53d624b9df93059/airflow/providers/amazon/aws/hooks/emr_containers.py#L174-L176

More information about the concerns can be found on https://github.com/apache/airflow/pull/16766#discussion_r668089559

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by victorphoenix3**
Query status polling logic currently polls until the query is not in intermmediate state. It should instead poll until query is not in the terminal state.
Resolves #19877

### CODE REVIEW

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (ferruzzi) [inline airflow/providers/amazon/aws/hooks/emr.py:121]: Is "CANCEL_PENDING" really a terminal state?

Looks good to me other than this question.

Comment (victorphoenix3) [reply airflow/providers/amazon/aws/hooks/emr.py:121]: "CANCEL_PENDING" is being treated as a [failure state](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/emr.py#L111-L115) for the EMR container sensor, which is why I included it as a terminal state. Should it be removed?

**Comment by victorphoenix3** [comment]
@ferruzzi A similar polling function exists for Athena queries as well [[link]](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/athena.py#L188). Should I modify that too?

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (ferruzzi) [reply airflow/providers/amazon/aws/hooks/emr.py:121]: I don't know the EMR behavior well enough to say.  If you feel it's intentional and the right thing to do, then leave it in.  Just asking because the name strikes me as an intermediate state.

**Comment by ferruzzi** [comment]
@victorphoenix3 Yeah, it looks like the exact same issue.  If you are up for it, then by all means!  But I'd say drop it in a fresh PR so you don't tie this one up.

File: airflow/providers/amazon/aws/hooks/emr.py
Comment (victorphoenix3) [reply airflow/providers/amazon/aws/hooks/emr.py:121]: Going by the code [here](https://github.com/apache/airflow/blob/main/airflow/providers/amazon/aws/hooks/emr.py#L111-L115) and [here](https://github.com/apache/airflow/blob/main/tests/providers/amazon/aws/sensors/test_emr_containers.py#L69-L72), I'll leave it in as a terminal state.

**Comment by ferruzzi** [comment]
I like it.   You'll have to wait for a Committer to approve and merge, but I don't see any issues.

**Comment by potiuk** [comment]
Is it possible to add a unit test for that one (and rebase it ? ) I am happy to approve it but some unit testing could prevent regressions.

**Comment by victorphoenix3** [comment]
@potiuk I have added the unit tests, excuse the delay. Please review.

**Comment by potiuk** [comment]
The failing tests were unrelated (I am fixing them in a separate branch and updated constraints to get rid of those errors). I rebased the PR to latest main to re-test it.

**Review by potiuk — APPROVED** [review]

**Pull request merged**
